tools/whipinto/src/rtspclient.rs

```diff
@@ -0,0 +1,417 @@
+        username, realm, nonce, uri, response
+    )
+}
+
+fn parse_sdp(sdp: &str) -> Result<(Option<String>, Option<String>, Codec)> {
+    let mut video_track = None;
+    let mut audio_track = None;
+    let mut video_codec = None;
+
+    for line in sdp.lines() {
+        if line.starts_with("m=video") {
+            for line in sdp.lines() {
+                if line.starts_with("a=control:") {
+                    video_track = Some(line.replace("a=control:", "").trim().to_string());
+                } else if line.starts_with("a=rtpmap:") {
+                    let parts: Vec<&str> = line.split_whitespace().collect();
+                    if parts.len() > 1 {
+                        if parts[1].contains("H264") {
+                            video_codec = Some(Codec::H264);
+                        } else if parts[1].contains("AV1") {
+                            video_codec = Some(Codec::AV1);
+                        } else if parts[1].contains("VP8") {
+                            video_codec = Some(Codec::Vp8);
+                        } else if parts[1].contains("VP9") {
+                            video_codec = Some(Codec::Vp9);
+                        } else if parts[1].contains("Opus") {
+                            video_codec = Some(Codec::Opus);
+                        } else if parts[1].contains("G722") {
+                            video_codec = Some(Codec::G722);
+                        }
+                    }
+                }
+                if video_track.is_some() && video_codec.is_some() {
+                    break;
+                }
+            }
+        }
+        if line.starts_with("m=audio") {
+            for line in sdp.lines() {
+                if line.starts_with("a=control:") {
+                    audio_track = Some(line.replace("a=control:", "").trim().to_string());
+                    break;
+                }
+            }
+        }
+    }
+
+    // Ok((video_track, audio_track, video_codec))
+    match video_codec {
+        Some(codec) => Ok((video_track, audio_track, codec)),
+        None => Err(anyhow!("No valid video codec found in SDP")),
+    }
+}
+
+pub async fn setup_rtsp_session(rtsp_url: &str) -> Result<(u16, Codec)> {
+    let mut session = RtspSession::new(rtsp_url).await?;
+
+    // Send OPTIONS
+    session.send_options().await?;
+
+    // Send DESCRIBE
+    let describe_response = session.send_describe().await?;
+
+    let sdp = describe_response
+        .lines()
+        .skip_while(|line| !line.is_empty())
+        .skip(1)
+        .collect::<Vec<_>>()
+        .join("\n");
+
+    info!("SDP content: \n{}", sdp);
+
+    let (video_track, audio_track, video_codec) = parse_sdp(&sdp)?;
+
+    if video_track.is_none() && audio_track.is_none() {
+        return Err(anyhow!("No tracks found in SDP"));
+    }
+
+    let rtp_port = pick_unused_port().ok_or_else(|| anyhow!("No available port found"))?;
```

Comment:
I suggest first bind port, and after return port

Look like, port is `0` is range allocation port. 
```
    let listener = tokio::net::TcpListener::bind("0.0.0.0:0")
        .await
        .unwrap();

// get binding address and port
    listener.local_addr()
```

* * *

For example: If `pick_unused_port` pick port = `12345`, There have a application is this computer use `12345`, This application just start.

Concurrent the port maybe conflict, Although this possibility is low

BTW: It's a little problem

`pick_unused_port` range port is 15000..25000
```rust
        let port = rng.gen_range(15000..25000);
```
https://github.com/Dentosal/portpicker-rs/blob/master/src/lib.rs#L57